### PR TITLE
Configure Plop to enable one-touch component scaffolding

### PR DESCRIPTION
**Pull request by unscsprt**
### Future iterations

- Automatically add export for new `core` components
- Figure out how to keep the `package.json` templates up to date automatically
- Validation of input when creating new components (e.g. valid HTML tag, proper component name capitalization, etc.)

### CODE REVIEW

File: plop-templates/package/babel.config.js [hunk 1]
```diff
@@ -0,0 +1 @@
+module.exports = require('../../babel.config')
```
Comment (broox9) [inline plop-templates/package/babel.config.js]: nit: I have to look and see what this was inheriting from. I guessed it was the root babel config. Not a big deal but a small comment stating that being the

File: plop-templates/core/Component.js.hbs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+} from './utils'
+
+const {{componentName}} = mapProps(({ align, ...props }) => ({
+  ...props
+}))(styled.{{htmlTag}}`
+  // Example: Use palette colors
+  color: ${getPaletteColor('base')};
+
+  // Advanced: Apply variation style along with custom styling from theme.componentStyles[component][variation]
+  ${applyVariations('Text')}
```
Comment (achoboter) [inline plop-templates/core/Component.js.hbs]: `Text` should be the components name

File: plop-templates/core/Story.js.hbs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import React from 'react'
+import { storiesOf } from '@storybook/react'
+import {{componentName}} from '../src/{{componentName}}'
+
+storiesOf('{{componentName}}', module).add('default', () => (
+    <{{componentName}} color='primary.base'>
```
Comment (achoboter) [inline plop-templates/core/Story.js.hbs]: `color` should only be `primary`, it will use the `${getPaletteColor('base')}` from the component so it will already be `primary.base`

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -76,6 +76,17 @@ update its snapshot:
 npm test -- -u
 ```
 
+### Creating new components
+
+We use [Plop](https://plopjs.com/) to scaffold new components rather than copy/pasting a boilerplate.
```
Comment (BeniCheni) [inline CONTRIBUTING.md]: English nit 😅 

```suggestion
We use [Plop](https://plopjs.com/) to scaffold new components rather than copying & pasting from a boilerplate.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -76,6 +76,17 @@ update its snapshot:
...
+The configuration allows you to create new core and package components.
+Follow the steps below to create a new component:
+
+1. Run `npm run plop` and follow the prompts
+1. Run `npm run prepare` to tell Lerna to bootstrap your new component folder
+1. If creating a core component, add an export for your new component to `packages/core/src/index.js`
+1. Start working!
```
Comment (BeniCheni) [inline CONTRIBUTING.md]: How come these are all step 1)? Also "Start working!" could be dropped in interest for brevity.

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -76,6 +76,17 @@ update its snapshot:
...
 
+### Creating new components
+
+We use [Plop](https://plopjs.com/) to scaffold new components rather than copy/pasting a boilerplate.
+The configuration allows you to create new core and package components.
+Follow the steps below to create a new component:
+
+1. Run `npm run plop` and follow the prompts
+1. Run `npm run prepare` to tell Lerna to bootstrap your new component folder
+1. If creating a core component, add an export for your new component to `packages/core/src/index.js`
```
Comment (BeniCheni) [inline CONTRIBUTING.md]:
```suggestion
3. If creating a core component, add an import & an export for the new core component in `packages/core/src/index.js` file
```

File: plop-templates/core/Component.js.hbs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+import {
+  mapProps,
+  applyVariations,
+  getPaletteColor
+} from './utils'
+
+const {{componentName}} = mapProps(({ align, ...props }) => ({
+  ...props
+}))(styled.{{htmlTag}}`
+  // Example: Use palette colors
```
Comment (BeniCheni) [inline plop-templates/core/Component.js.hbs]: ✂️ comment?

File: plop-templates/core/Component.js.hbs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  color: ${getPaletteColor('base')};
+
+  // Advanced: Apply variation style along with custom styling from theme.componentStyles[component][variation]
```
Comment (BeniCheni) [inline plop-templates/core/Component.js.hbs]: I am all for comments to communicate, but if you choose to comment on here, I think it needs to be more expressive in plain English what it's doing?

Or else, we could ✂️ for brevity. I'd vote for ✂️comment, and encourage the code to rely on naming convention to indicate what the logic is.

File: plop-templates/package/rollup.config.js [hunk 9]
```diff
@@ -1,3 +1,4 @@
+// This refers to the Rollup config at the root of design-system
```
Comment (BeniCheni) [inline plop-templates/package/rollup.config.js]: ✂️ comment?

File: plop-templates/core/Component.js.hbs [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  ${applyVariations('{{componentName}}')}
+`)
+
+{{componentName}}.displayName = '{{componentName}}'
+
+{{componentName}}.propTypes = {
+}
+
+export default {{componentName}}
```
Comment (BeniCheni) [inline plop-templates/core/Component.js.hbs]: Overall I wonder if the name `componentName` here and in other places (files) could be just `component` instead?

File: CONTRIBUTING.md (hunk 5 above)
Comment (unscsprt) [reply CONTRIBUTING.md]: That's a way to do ordered lists in MD. I'll remove the last part as suggested.

File: plop-templates/core/Component.js.hbs (hunk 10 above)
Comment (unscsprt) [reply plop-templates/core/Component.js.hbs]: I kinda wanted to make it obvious (even though it is an HBS template) that the thing we're inserting is a name

File: plop-templates/package/rollup.config.js (hunk 9 above)
Comment (unscsprt) [reply plop-templates/package/rollup.config.js]: I _think_ that was a @sdalonzo suggestion; I think it'd be okay to keep just to be clear where its rollup config values are coming from, where it's shared, etc.

File: package.json [hunk 11]
```diff
@@ -11,7 +11,8 @@
     "coverage": "jest --coverage",
     "codecov": "npx codecov",
     "start": "start-storybook -p 8000 -c storybook",
-    "storybook": "build-storybook -c storybook -o docs/out/storybook"
+    "storybook": "build-storybook -c storybook -o docs/out/storybook",
+    "plop": "npm run plop"
```
Comment (sdalonzo) [inline package.json]: perhaps the script should have a more semantic name like `create-component` or something similar

File: plop-templates/package/.gitignore [hunk 12]
```diff
@@ -0,0 +1,18 @@
+node_modules
```
Comment (sdalonzo) [inline plop-templates/package/.gitignore]: add the 2 new entries from the rollup PR: 
`stats.html`
`rollup.bundleAnalysis.json`

**Comment by BeniCheni** [comment]
@unscsprt mostly LGTM. Plz rebase and resolve conflicts to become mergeable to land ✅.

File: plop-templates/core/Story.js.hbs (hunk 3 above)
Comment (unscsprt) [reply plop-templates/core/Story.js.hbs]: I forget what the exact reason was, but @bertya gave an explanation yesterday as to why keeping `base` is preferable.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -76,6 +76,16 @@ update its snapshot:
 npm test -- -u
 ```
 
+### Creating new components
+
+We use [Plop](https://plopjs.com/) to scaffold new components rather than copying & pasting from a boilerplate.
+The configuration allows you to create new core and package components.
+Follow the steps below to create a new component:
+
+1. Run `npm run plop` and follow the prompts
```
Comment (sdalonzo) [inline CONTRIBUTING.md]: update this to match the `create-component` script
